parser/util/Environment.js

```diff
@@ -0,0 +1,35 @@
+import { LinkedList } from './LinkedList';
+
+/**
+ * A special key-value pair linked list that can lookup values of a certain key.
+ */
+export class Environment extends LinkedList {
+  constructor (head, tail) {
+    super(head, tail);
+    let hed = this.head;
+    this.head = (value) => value === undefined ? hed() : hed([hed()[0], value]);
```

Comment:
Refactor to be more readable. Same with my comment with the `LinkedList.head()` details.